extensions/ql-vscode/src/test-manager.ts

```diff
@@ -0,0 +1,373 @@
+      const workspaceFolderHandler =
+        this.workspaceFolderHandlers.get(workspaceFolder);
+      if (workspaceFolderHandler !== undefined) {
+        // Delete the root item for this workspace folder, if any.
+        this.testController.items.delete(workspaceFolder.uri.toString());
+        this.disposeAndStopTracking(workspaceFolderHandler);
+        this.workspaceFolderHandlers.delete(workspaceFolder);
+      }
+    }
+  }
+
+  private handleDidChangeWorkspaceFolders(
+    e: WorkspaceFoldersChangeEvent,
+  ): void {
+    this.startTrackingWorkspaceFolders(e.added);
+    this.stopTrackingWorkspaceFolders(e.removed);
+  }
+
+  /**
+   * Update the test controller when we discover changes to the tests in the workspace folder.
+   */
+  public updateTestsForWorkspaceFolder(
+    workspaceFolder: WorkspaceFolder,
+    testDirectory: QLTestDirectory | undefined,
+  ): void {
+    if (testDirectory !== undefined) {
+      // Adding an item with the same ID as an existing item will replace it, which is exactly what
+      // we want.
+      // Test discovery returns a root `QLTestDirectory` representing the workspace folder itself,
+      // named after the `WorkspaceFolder` object's `name` property. We can map this directly to a
+      // `TestItem`.
+      this.testController.items.add(
+        this.createTestItemTree(testDirectory, true),
+      );
+    } else {
+      // No tests, so delete any existing item.
+      this.testController.items.delete(workspaceFolder.uri.toString());
+    }
+  }
+
+  /**
+   * Creates a tree of `TestItem`s from the root `QlTestNode` provided by test discovery.
+   */
+  private createTestItemTree(node: QLTestNode, isRoot: boolean): TestItem {
+    // Prefix the ID to identify it as a directory or a test
+    const itemType = node instanceof QLTestDirectory ? "dir" : "test";
+    const testItem = this.testController.createTestItem(
+      // For the root of a workspace folder, use the full path as the ID. Otherwise, use the node's
+      // name as the ID, since it's shorter but still unique.
+      `${itemType} ${isRoot ? node.path : node.name}`,
+      node.name,
+      Uri.file(node.path),
+    );
+
+    for (const childNode of node.children) {
+      const childItem = this.createTestItemTree(childNode, false);
+      if (childNode instanceof QLTestFile) {
+        childItem.range = new Range(0, 0, 0, 0);
+      }
+      testItem.children.add(childItem);
+    }
+
+    return testItem;
+  }
+
+  /**
+   * Run the tests specified by the `TestRunRequest` parameter.
+   *
+   * Public because this is used in unit tests.
+   */
+  public async run(
+    request: TestRunRequest,
+    token: CancellationToken,
+  ): Promise<void> {
+    const testsToRun = this.computeTestsToRun(request);
+    const testRun = this.testController.createTestRun(request, undefined, true);
+    try {
+      const tests: string[] = [];
+      testsToRun.forEach((t) => {
+        testRun.enqueued(t);
+        tests.push(t.uri!.fsPath);
```

Comment:
Can we do something different here, like returning if `t.uri === undefined`?

Reply:
We actually already had the path we need in the key of the map entry, so I used that instead.